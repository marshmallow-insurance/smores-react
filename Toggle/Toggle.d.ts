import { FC } from 'react';
import { MarginProps } from '../utils/space';
type Props = {
    /** unique ID */
    id?: string;
    /** checked flag  */
    checked: boolean;
    /** onToggle listener  */
    onToggle: () => void;
} & MarginProps;
export declare const Toggle: FC<Props>;
export {};
