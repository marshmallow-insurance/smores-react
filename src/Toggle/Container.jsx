import React, {useState} from 'react';
import {Toggle} from './Toggle';

export const Container = () => {
  const [active, setActive] = useState(false);

  return <Toggle checked={active} onToggle={() => setActive(!active)} />;
};
