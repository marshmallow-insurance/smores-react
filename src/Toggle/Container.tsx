import React, { useState } from 'react'
import { Toggle } from './Toggle'

export const Container = () => {
  const [active, setActive] = useState(false)

  return (
    <Toggle
      id="toggleId"
      checked={active}
      onToggle={() => setActive(!active)}
    />
  )
}
