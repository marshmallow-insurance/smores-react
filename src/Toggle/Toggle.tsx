import React, { FC } from 'react'
import styled from 'styled-components'

import { focusOutline } from '../utils/focusOutline'
import { theme } from '../theme'

type Props = {
  /** unique ID */
  id?: string
  /** checked flag  */
  checked: boolean
  /** onToggle listener  */
  onToggle: () => void
}

export const Toggle: FC<Props> = ({ checked, onToggle }) => {
  return (
    <Switch>
      <Checkbox type="checkbox" checked={checked} onChange={onToggle} />
      <Slider />
    </Switch>
  )
}

const Switch = styled.label`
  position: relative;
  display: inline-block;
  width: 50px;
  height: 32px;

  input {
    opacity: 0;
    width: 0;
    height: 0;
  }
`

const Slider = styled.span`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${theme.colors.bgPrimary};
  border: 1px solid;
  border-color: ${theme.colors.bgPrimary};
  border-radius: 28px;
  transition: 0.2s background-color border-color;
  outline: none;

  &:before {
    position: absolute;
    content: '';
    height: 24px;
    width: 24px;
    left: 3px;
    bottom: 3px;
    background-color: ${theme.colors.white};
    transition: 0.2s transform;
    border-radius: 50%;
  }
`

const Checkbox = styled.input`
  ${focusOutline({ selector: `&:focus-visible + ${Slider}` })}

  &:checked + ${Slider} {
    background-color: ${theme.colors.success};
    border: 1px solid;
    border-color: ${theme.colors.success};
  }

  &:checked + ${Slider}:before {
    transform: translateX(18px);
  }
`
