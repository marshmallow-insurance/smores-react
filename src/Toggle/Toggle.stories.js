import React from 'react'
import { storiesOf } from '@storybook/react'

import { Container } from './Container.jsx'

storiesOf('Toggle', module).addWithJSX('default', () => <Container />)
