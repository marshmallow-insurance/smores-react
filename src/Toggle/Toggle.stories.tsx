import React from 'react'
import { Toggle } from './Toggle'
import { Container } from './Container'

export default {
  title: 'Toggle',
  component: Toggle,
}

const Template = () => <Container />

export const Default = Template.bind({})
