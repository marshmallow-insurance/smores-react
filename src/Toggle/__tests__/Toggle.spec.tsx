import { render } from '@testing-library/react'
import React from 'react'

import { Toggle } from '../Toggle'

test('renders', () => {
  const { container } = render(
    <Toggle id="toggle" checked={false} onToggle={() => {}} />,
  )
  expect(container.firstChild).toMatchSnapshot()
})
