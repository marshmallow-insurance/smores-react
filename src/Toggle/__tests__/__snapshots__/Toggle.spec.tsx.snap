// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<label
  class="sc-aXZVg sc-gEvEer byxNQM jwXIYA"
  id="toggle"
>
  <input
    class="sc-fqkvVR egoFiG"
    type="checkbox"
  />
  <span
    class="sc-eqUAAy kvquKw"
  />
</label>
`;
